### PR TITLE
Module4

### PR DESCRIPTION
**Pull request by Katerina-Charakhovich**
Module 4

### CODE REVIEW

File: model/src/main/java/com/epam/esm/model/CustomPage.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+@Data
+public class CustomPage {
+    private static final int MIN_PAGE_NUMBER = 0;
+    private static final int MAX_PAGE_SIZE = 5;
+    private static final int MIN_PAGE_SIZE = 1;
+    private int pageNumber;
+    private int pageSize;
+    private int pageTotal;
+
+    public CustomPage(int pageNumber,int  pageSize,Long pageTotal) {
```
Comment (valevaty98) [inline model/src/main/java/com/epam/esm/model/CustomPage.java]: This logic is difficult to perceive,
also not auto-formatted

File: web/src/main/java/com/epam/esm/web/config/ResourceServerConfig.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    public void configure(HttpSecurity http) throws Exception {
+        http.httpBasic().disable().csrf().disable().sessionManagement()
+                .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and().authorizeRequests()
+                .antMatchers(HttpMethod.POST, REGISTER).permitAll()
+                .antMatchers(HttpMethod.GET, CERTIFICATES).permitAll()
+                .antMatchers(HttpMethod.GET, TAGS).permitAll()
+                .anyRequest().authenticated().and()
+                .cors().and().exceptionHandling().and()
+                .csrf().disable();
```
Comment (valevaty98) [inline web/src/main/java/com/epam/esm/web/config/ResourceServerConfig.java]: you already disabled csrf at line 29

File: web/src/main/java/com/epam/esm/web/controller/PurchaseController.java [hunk 3]
```diff
@@ -47,28 +42,29 @@ public PurchaseController(PurchaseService purchaseService, HateoasWrapper hateoa
     }
 
     /**
-     * Find all users
+     * Find all purchases
      *
      * @return list
      */
+    @PreAuthorize("isAuthenticated()")
```
Comment (valevaty98) [inline web/src/main/java/com/epam/esm/web/controller/PurchaseController.java]: As far as I see in your approach, every user can access the purchases of other users. This is not entirely correct. 
A user should have an access only to the purchases he owns. The only exclusion is an admin who can see the purchases of other users.

File: web/src/main/java/com/epam/esm/web/controller/UserController.java [hunk 4]
```diff
@@ -69,12 +65,11 @@ public UserController(UserService userService, PurchaseService purchaseService,
      */
     @GetMapping
     public ResponseEntity<List<UserDto>> findAll(
```
Comment (valevaty98) [inline web/src/main/java/com/epam/esm/web/controller/UserController.java:77]: This endpoint also should be secured

File: web/src/main/java/com/epam/esm/web/controller/UserController.java [hunk 5]
```diff
@@ -83,23 +78,32 @@ public UserController(UserService userService, PurchaseService purchaseService,
      *
      * @return list
      */
+    @PreAuthorize("isAuthenticated()")
```
Comment (valevaty98) [inline web/src/main/java/com/epam/esm/web/controller/UserController.java]: Again, according to this, every user can see purchases of another user only by specifying the id of this another user.
This private info should be accessible only be the owner of this info or an admin

File: web/src/main/java/com/epam/esm/web/controller/UserController.java [hunk 6]
```diff
@@ -83,23 +78,32 @@ public UserController(UserService userService, PurchaseService purchaseService,
...
     }
 
     /**
      * Create purchase for user
      *
      * @return PurchaseShortDto
      */
+    @PreAuthorize("hasAuthority('ROLE_USER')")
     @PostMapping("/{id}/purchases")
     public ResponseEntity<PurchaseShortDto> addPurchaseByUserId(@Valid @RequestBody PurchaseShortDto purchaseCreateDto,
```
Comment (valevaty98) [inline web/src/main/java/com/epam/esm/web/controller/UserController.java:111]: According to this, any user can add a purchase to any other user.
ONLY user should be authorized to acquire certificates for himself

File: web/src/main/java/com/epam/esm/web/config/ResourceServerConfig.java (hunk 2 above)
Comment (Katerina-Charakhovich) [reply web/src/main/java/com/epam/esm/web/config/ResourceServerConfig.java]: done

File: web/src/main/java/com/epam/esm/web/controller/UserController.java (hunk 4 above)
Comment (Katerina-Charakhovich) [reply web/src/main/java/com/epam/esm/web/controller/UserController.java:77]: done

File: web/src/main/java/com/epam/esm/web/controller/UserController.java (hunk 6 above)
Comment (Katerina-Charakhovich) [reply web/src/main/java/com/epam/esm/web/controller/UserController.java:111]: done

File: web/src/main/java/com/epam/esm/web/controller/PurchaseController.java (hunk 3 above)
Comment (Katerina-Charakhovich) [reply web/src/main/java/com/epam/esm/web/controller/PurchaseController.java]: done

**Review by valevaty98 — APPROVED** [review]

File: model/src/main/java/com/epam/esm/model/CustomPage.java (hunk 1 above)
Comment (Katerina-Charakhovich) [reply model/src/main/java/com/epam/esm/model/CustomPage.java]: done

**Pull request merged**